### ISSUE TITLE
[PT2D] Option to display TORCH_LOGS from one rank

### ISSUE THREAD

**Issue by xmfan** [pytorch/pytorch#117383]
### 🚀 The feature, motivation and pitch

We use TORCH_LOGS often for graph inspection, but under distributed launcher, it currently displays interleaved from every rank. This makes it difficult to read sequential logs like dynamo TRACE.

We should consider a way to only show TORCH_LOGS from certain ranks e.g. TORCH_LOGS_RANKS=0,1. syntax similar to CUDA_VISIBLE_DEVICES

repro
`TORCH_LOGS=+dynamo torchrun --nproc_per_node=8 t.py`
```
[rank1]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE CALL_FUNCTION 0 [UserMethodVariable(<function DistributedDataParallel._lazy_init at 0x7f5c5600b880>, UnspecializedNNModuleVariable(DistributedDataParallel))]                                                                                          
[rank1]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert: [DEBUG] INLINING <code object _lazy_init at 0x7f5c56006600, file "/home/xmfan/core/pytorch/torch/nn/parallel/distributed.py", line 1382>, inlined according skipfiles.MOD_INLINELIST                                                                                 
[rank6]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.convert_frame: [INFO] Restarting analysis due to _dynamo/symbolic_convert.py:141 in fail_and_restart_analysis                                                                                                                                                                          
[rank1]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG] TRACE starts_line /home/xmfan/core/pytorch/torch/nn/parallel/distributed.py:1382 in _lazy_init (DistributedDataParallel._lazy_init) (inline depth: 4)                                                                                         
[rank1]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG]         def _lazy_init(self):                                                                                                                                                                                                                 
[rank4]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG] TRACE starts_line /home/xmfan/core/pytorch/torch/nn/parallel/distributed.py:917 in _setup_in_backward_optimizers (DistributedDataParallel._setup_in_backward_optimizers) (inline depth: 5)                                                    
[rank4]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG]         def _setup_in_backward_optimizers(self):                                                                                                                                                                                              
[rank1]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG] TRACE starts_line /home/xmfan/core/pytorch/torch/nn/parallel/distributed.py:1385 in _lazy_init (DistributedDataParallel._lazy_init) (inline depth: 4)                                                                                         
[rank1]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG]             self._setup_in_backward_optimizers()                                                                                                                                                                                              
[rank1]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST self []                                                                                                                                                                                                                                      
[rank4]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG] TRACE starts_line /home/xmfan/core/pytorch/torch/nn/parallel/distributed.py:928 in _setup_in_backward_optimizers (DistributedDataParallel._setup_in_backward_optimizers) (inline depth: 5)                                                    
[rank4]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG]             if any(hasattr(p, "_in_backward_optimizers") for p in self._module_parameters):                                                                                                                                                   
[rank1]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_ATTR _setup_in_backward_optimizers [UnspecializedNNModuleVariable(DistributedDataParallel)]                                                                                                                                                       
[rank4]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_GLOBAL any []                                                                                                                                                                                                                                     
[rank6]:[2024-01-16 22:29:36,221] [0/0_1] torch._dynamo.symbolic_convert: [INFO] Step 1: torchdynamo start tracing inner /home/xmfan/core/pytorch/torch/_dynamo/external_utils.py:23                                                                                                                                                         
[rank4]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_CONST <code object <genexpr> at 0x7fa69a7fcf50, file "/home/xmfan/core/pytorch/torch/nn/parallel/distributed.py", line 928> [BuiltinVariable(any)]                                                                                                
[rank6]:[2024-01-16 22:29:36,221] [0/0_1] torch.fx.experimental.symbolic_shapes: [INFO] create_env                                                                                                                                                                                                                                           
[rank4]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_CONST DistributedDataParallel._setup_in_backward_optimizers.<locals>.<genexpr> [BuiltinVariable(any), ConstantVariable(code)]                                                                                                                     
[rank4]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE MAKE_FUNCTION 0 [BuiltinVariable(any), ConstantVariable(code), ConstantVariable(str)]                                                                                                                                                                  
[rank1]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE CALL_FUNCTION 0 [UserMethodVariable(<function DistributedDataParallel._setup_in_backward_optimizers at 0x7f5c5600acb0>, UnspecializedNNModuleVariable(DistributedDataParallel))]                                                                       
[rank4]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST self [BuiltinVariable(any), NestedUserFunctionVariable()]                                                                                                                                                                                    
[rank4]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_ATTR _module_parameters [BuiltinVariable(any), NestedUserFunctionVariable(), UnspecializedNNModuleVariable(DistributedDataParallel)]                                                                                                              
[rank1]:[2024-01-16 22:29:36,221] [0/0] torch._dynamo.symbolic_convert: [DEBUG] INLINING <code object _setup_in_backward_optimizers at 0x7f5c56005000, file "/home/xmfan/core/pytorch/torch/nn/parallel/distributed.py", line 917>, inlined according skipfiles.MOD_INLINELIST                                                               
[rank6]:[2024-01-16 22:29:36,221] [0/0_1] torch._dynamo.symbolic_convert.__trace_source: [DEBUG] TRACE starts_line /home/xmfan/core/pytorch/torch/_dynamo/external_utils.py:23 in inner (wrap_inline)                                                                                                                                        
[rank6]:[2024-01-16 22:29:36,221] [0/0_1] torch._dynamo.symbolic_convert.__trace_source: [DEBUG]         @functools.wraps(fn)                                                                                                                                                                                                                
[rank6]:[2024-01-16 22:29:36,222] [0/0_1] torch._dynamo.symbolic_convert.__trace_source: [DEBUG] TRACE starts_line /home/xmfan/core/pytorch/torch/_dynamo/external_utils.py:25 in inner (wrap_inline.inner)                                                                                                                                  
[rank6]:[2024-01-16 22:29:36,222] [0/0_1] torch._dynamo.symbolic_convert.__trace_source: [DEBUG]             return fn(*args, **kwargs)                                                                                                                                                                                                      
[rank4]:[2024-01-16 22:29:36,222] [0/0] torch._dynamo.output_graph: [DEBUG] create_graph_input L_fn_module_parameters_0_ L['fn']._module_parameters[0]                                                                                                                                                                                       
[rank6]:[2024-01-16 22:29:36,222] [0/0_1] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_DEREF fn []                                                                                                                                                                                                                                     
[rank6]:[2024-01-16 22:29:36,222] [0/0_1] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST args [LazyVariableTracker()]                                                                                                                                                                                                               
[rank6]:[2024-01-16 22:29:36,222] [0/0_1] torch._dynamo.symbolic_convert: [DEBUG] TRACE BUILD_MAP 0 [LazyVariableTracker(), LazyVariableTracker()]                                                                                                                                                                                           
[rank1]:[2024-01-16 22:29:36,222] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG] TRACE starts_line /home/xmfan/core/pytorch/torch/nn/parallel/distributed.py:917 in _setup_in_backward_optimizers (DistributedDataParallel._setup_in_backward_optimizers) (inline depth: 5)                                                    [rank1]:[2024-01-16 22:29:36,222] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG]         def _setup_in_backward_optimizers(self):                                                                                                                                                                                              [rank6]:[2024-01-16 22:29:36,222] [0/0_1] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST kwargs [LazyVariableTracker(), LazyVariableTracker(), ConstDictVariable()]                                                                                                                                                                 
[rank1]:[2024-01-16 22:29:36,222] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG] TRACE starts_line /home/xmfan/core/pytorch/torch/nn/parallel/distributed.py:928 in _setup_in_backward_optimizers (DistributedDataParallel._setup_in_backward_optimizers) (inline depth: 5)                                                    
[rank1]:[2024-01-16 22:29:36,222] [0/0] torch._dynamo.symbolic_convert.__trace_source: [DEBUG]             if any(hasattr(p, "_in_backward_optimizers") for p in self._module_p
...
```

### Alternatives

There is already a wider effort to get better distributed failure messages, maybe the implementation can also cover TORCH_LOGS

### Additional context

_No response_

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519

**Comment by ezyang** [comment]
By default, PT2 is supposed to generate logs that include rank information, see:

```
        record.rankprefix = ""
        if dist.is_available() and dist.is_initialized():
            record.rankprefix = f"[rank{dist.get_rank()}]:"
```

in `torch/_logging/_internal.py`.

If you are not seeing this, that means someone is overriding the logging format. Can torchrun stop overriding the logging format instead?

**Comment by wconstab** [comment]
cc @kurman are we overriding log format in torchrun?

**Comment by xmfan** [comment]
> If you are not seeing this, that means someone is overriding the logging format. Can torchrun stop overriding the logging format instead?

I had an `export TORCH_LOGS_FORMAT="%(levelname)s: %(message)s"` in my bashrc. This was hiding the rank prefix, torchrun isn't overriding anything, updated the description.

The issue with poor readability of interleaved logs still remains.

**Comment by wconstab** [comment]
I think it's a nice idea to provide a rank filter that integrates with torch_logs.  

Something annoying might be that cpp logs won't honor it at least without more work.

We also had a task in the same parent issue tracker for putting a mutex around writing so that we don't interleave parts of log strings, which might also help your case.

**Comment by drisspg** [comment]
+1

**Comment by xmfan** [comment]
I'm rethinking the UX. Integrating into the python torch logger limits the filtering for only python logs, and also assumes the distributed process group's shape before it is initialized.

After learning about `torchrun --role process`, it seems like the better module to integrate with. It can already prefix each log entry with the rank (even c++ logs) e.g. `TORCH_LOGS="+dynamo" torchrun --role rank --nnodes 1 --nproc-per-node 8 t.py`. I can add a rank filtering option for stdout/stderr.

**Comment by wconstab** [comment]
what does the --role argument do and how is that related to your log proposal? 
from torchrun's help: `  --role ROLE           User-defined role for the workers.`

**Comment by xmfan** [comment]
torchrun --role will assign a string and an id to each process it launches. From my usage, I have only seen it used to prefix logs with rank information e.g. `--role rank` ==> `[rankX]` where X in [0, local ranks[

> After learning about `torchrun --role process` 

To clarify, I meant to say something along the lines of: Since torchrun --role can prefix even c++ logs with rank information, I spent some time looking at the implementation and see that it is using the multiprocess/subprocess APIs which allow specifying a stream for stdout/stderr. This would be a better place to filter logs, since it would include all logs for the rank, regardless of whether it comes from python, c++ or even pytorch". I implemented it in https://github.com/pytorch/pytorch/pull/118562
